drivers/pwm/pwm_mspm0.c

```diff
@@ -0,0 +1,168 @@
+	const uint32_t frequency;
+	const DL_Timer_ClockConfig clock_cfg;
+	const DL_TIMER_PWM_MODE pwm_mode;
+};
+
+static inline int pwm_mspm0_get_cc_direction(int cc_index, int *direction)
+{
+	switch (cc_index) {
+	case DL_TIMER_CC_0_INDEX:
+		*direction = DL_TIMER_CC0_OUTPUT;
+		return 0;
+	case DL_TIMER_CC_1_INDEX:
+		*direction = DL_TIMER_CC1_OUTPUT;
+		return 0;
+	case DL_TIMER_CC_2_INDEX:
+		*direction = DL_TIMER_CC2_OUTPUT;
+		return 0;
+	case DL_TIMER_CC_3_INDEX:
+		*direction = DL_TIMER_CC3_OUTPUT;
+		return 0;
+	default:
+		LOG_DBG("Error, invalid CC index: %d", cc_index);
+		return -EINVAL;
+	};
+}
+
+static int pwm_mspm0_get_cycles_per_sec(const struct device *dev, uint32_t channel,
+					uint64_t *cycles)
+{
+	const struct pwm_mspm0_config *config = dev->config;
+
+	*cycles = config->frequency /
+		  ((config->clock_cfg.divideRatio + 1) * (config->clock_cfg.prescale + 1));
+
+	return 0;
+}
+
+static inline uint32_t get_polarity(pwm_flags_t flags)
+{
+	if ((flags & PWM_POLARITY_MASK) == PWM_POLARITY_NORMAL) {
+		return DL_TIMER_CC_OCTL_INV_OUT_DISABLED;
+	}
+
+	return DL_TIMER_CC_OCTL_INV_OUT_ENABLED;
+}
+
+static int pwm_mspm0_set_cycles(const struct device *dev, uint32_t channel, uint32_t period_cycles,
+				uint32_t pulse_cycles, pwm_flags_t flags)
+{
+	int direction;
+	const struct pwm_mspm0_config *cfg = dev->config;
+	uint8_t max_allowed_channels = HAS_FOUR_CC(cfg->timer) ? MAX_ALLOWED_CHANNELS_WITH_FOUR_CC
+							       : MAX_ALLOWED_CHANNELS_TIMX;
+
+	if (channel >= max_allowed_channels) {
+		LOG_ERR("Error, invalid channel: %d", channel);
+		return -EINVAL;
+	}
+
+	if (pulse_cycles == 0) {
+		DL_Timer_stopCounter(cfg->timer);
+		return 0;
+	}
+
+	if (IS_TIMA(cfg->timer)) {
+		const DL_TimerA_PWMConfig pwm_cfg = {
+			.pwmMode = cfg->pwm_mode,
+			.isTimerWithFourCC = HAS_FOUR_CC(cfg->timer),
+			.period = period_cycles,
+		};
+
+		DL_TimerA_initPWMMode(cfg->timer, (DL_TimerA_PWMConfig *)&pwm_cfg);
+	} else {
+		const DL_TimerG_PWMConfig pwm_cfg = {
+			.pwmMode = cfg->pwm_mode,
+			.period = period_cycles,
+		};
+
+		DL_TimerG_initPWMMode(cfg->timer, (DL_TimerG_PWMConfig *)&pwm_cfg);
+	}
+
+	DL_Timer_setCaptureCompareOutCtl(cfg->timer, DL_TIMER_CC_OCTL_INIT_VAL_HIGH,
+					 get_polarity(flags), DL_TIMER_CC_OCTL_SRC_FUNCVAL,
+					 channel);
+	int err = pwm_mspm0_get_cc_direction(channel, &direction);
+	if (err) {
+		DL_Timer_disablePower(cfg->timer);
+		return err;
+	}
+
+	DL_Timer_setCaptureCompareValue(cfg->timer, pulse_cycles, channel);
+	DL_Timer_setCCPDirection(cfg->timer, direction);
+	DL_Timer_startCounter(cfg->timer);
+
+	return 0;
+}
+
+static const struct pwm_driver_api pwm_mspm0_driver_api = {
+	.set_cycles = pwm_mspm0_set_cycles,
+	.get_cycles_per_sec = pwm_mspm0_get_cycles_per_sec,
+};
+
+static int pwm_mspm0_init(const struct device *dev)
+{
+	const struct pwm_mspm0_config *cfg = dev->config;
+
+	int ret = pinctrl_apply_state(cfg->pinctrl, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
```

Comment:
```suggestion
	if (!ret) {
```
Checking specifically for negative values indicate to the reader that a positive value is normal and OK. But pinctrl_apply_state returns `0` on success, so I'd tighten the check to specifically accept only `0`

Reply:
i think you meant `if (ret)`